File: Training #2/C++/Juanes/friends.cpp
#include <iostream>
#include <vector>
#include <cstdio>

using namespace std;

int main() {
    int test_cases;
    scanf("%d", &test_cases);

    while (test_cases--) {
        int town_citizens, pairs_of_people;
        scanf("%d %d", &town_citizens, &pairs_of_people);

        vector<int> disSet(town_citizens, -1);


        for (int i = 0; i < pairs_of_people; i++) {
            int citizen1, citizen2;
            scanf("%d %d", &citizen1, &citizen2);

            citizen1--;
            citizen2--;

            while (disSet[citizen1] >= 0) {
                citizen1 = disSet[citizen1];
            }
            while (disSet[citizen2] >= 0) {
                citizen2 = disSet[citizen2];
            }

            if (citizen1 != citizen2) {
                if (citizen1 > citizen2) {
                    swap(citizen1, citizen2);
                }
                disSet[citizen1] += disSet[citizen2];
                disSet[citizen2] = citizen1;
            }
        }

        int max_size = 0;
        for (int i = 0; i < town_citizens; i++) {
            if (disSet[i] < 0) {
                max_size = max(max_size, -disSet[i]);
            }
        }

        printf("%d\n", max_size);
    }

    return 0;
}

File: Trainings/Training #2/Python/Juanes/friends.py
from sys import stdin


def main():
    test_cases = int(stdin.readline().strip())

    for _ in range(test_cases):

        town_citizens, pairs_of_people = map(int, stdin.readline().strip().split())


        disSet = [-1] * town_citizens

        for _ in range(pairs_of_people):
            citizen1, citizen2 = map(int, stdin.readline().strip().split())

            citizen1 -= 1
            citizen2 -= 1


            while disSet[citizen1] >= 0:
                citizen1 = disSet[citizen1]

            while disSet[citizen2] >= 0:
                citizen2 = disSet[citizen2]

            if citizen1 == citizen2:
                continue

            if citizen1 > citizen2:
                citizen1, citizen2 = citizen2, citizen1
            disSet[citizen1] += disSet[citizen2]
            disSet[citizen2] = citizen1

        print(-min(disSet))

if __name__ == "__main__":
    main()